Mathlib/Probability/Moments/SubGaussian.lean

```diff
@@ -0,0 +1,386 @@
+    MemLp (fun ω ↦ exp (t * X ω)) p (κ ∘ₘ ν) := by
+  by_cases hp0 : p = 0
+  · simpa [hp0] using (h.integrable_exp_mul t).1
+  constructor
+  · exact (h.integrable_exp_mul t).1
+  · rw [eLpNorm_lt_top_iff_lintegral_rpow_enorm_lt_top (mod_cast hp0) (by simp)]
+    simp only [ENNReal.coe_toReal]
+    have h' := (h.integrable_exp_mul (p * t)).2
+    rw [hasFiniteIntegral_def] at h'
+    convert h' using 3 with ω
+    rw [enorm_eq_ofReal (by positivity), enorm_eq_ofReal (by positivity),
+      ENNReal.ofReal_rpow_of_nonneg (by positivity), ← exp_mul, mul_comm, ← mul_assoc]
+    positivity
+
+lemma cgf_le (h : HasSubgaussianMGF X c κ ν) :
+    ∀ᵐ ω' ∂ν, ∀ t, cgf X (κ ω') t ≤ c * t ^ 2 / 2 := by
+  filter_upwards [h.mgf_le, h.ae_forall_integrable_exp_mul] with ω' h h_int t
+  calc cgf X (κ ω') t
+  _ = log (mgf X (κ ω') t) := rfl
+  _ ≤ log (exp (c * t ^ 2 / 2)) := by
+    by_cases h0 : κ ω' = 0
+    · simpa [h0] using by positivity
+    gcongr
+    · exact mgf_pos' h0 (h_int t)
+    · exact h t
+  _ ≤ c * t ^ 2 / 2 := by rw [log_exp]
+
+end BasicProperties
+
+protected lemma of_rat (h_int : ∀ t : ℝ, Integrable (fun ω ↦ exp (t * X ω)) (κ ∘ₘ ν))
+    (h_mgf : ∀ q : ℚ, ∀ᵐ ω' ∂ν, mgf X (κ ω') q ≤ exp (c * q ^ 2 / 2)) :
+    Kernel.HasSubgaussianMGF X c κ ν where
+  integrable_exp_mul := h_int
+  mgf_le := by
+    rw [← ae_all_iff] at h_mgf
+    have h_int : ∀ᵐ ω' ∂ν, ∀ t, Integrable (fun ω ↦ exp (t * X ω)) (κ ω') := by
+      have h_int' (n : ℤ) := Measure.ae_integrable_of_integrable_comp (h_int n)
+      rw [← ae_all_iff] at h_int'
+      filter_upwards [h_int'] with ω' h_int t
+      exact integrable_exp_mul_of_le_of_le (h_int _) (h_int _) (Int.floor_le t) (Int.le_ceil t)
+    filter_upwards [h_mgf, h_int] with ω' h_mgf h_int t
+    refine Rat.denseRange_cast.induction_on t ?_ h_mgf
+    exact isClosed_le (continuous_mgf h_int) (by fun_prop)
+
+@[simp]
+lemma fun_zero [IsFiniteMeasure ν] [IsZeroOrMarkovKernel κ] :
+    HasSubgaussianMGF (fun _ ↦ 0) 0 κ ν where
+  integrable_exp_mul := by simp
+  mgf_le := by simpa using ae_of_all _ fun _ ↦ toReal_prob_le_one
+
+@[simp]
+lemma zero [IsFiniteMeasure ν] [IsZeroOrMarkovKernel κ] : HasSubgaussianMGF 0 0 κ ν := fun_zero
+
+lemma congr {Y : Ω → ℝ} (h : HasSubgaussianMGF X c κ ν) (h' : X =ᵐ[κ ∘ₘ ν] Y) :
+    HasSubgaussianMGF Y c κ ν where
+  integrable_exp_mul t := by
+    refine (integrable_congr ?_).mpr (h.integrable_exp_mul t)
+    filter_upwards [h'] with ω hω using by rw [hω]
+  mgf_le := by
+    have h'' := Measure.ae_ae_of_ae_comp h'
+    filter_upwards [h.mgf_le, h''] with ω' h_mgf h' t
+    rw [mgf_congr (Filter.EventuallyEq.symm h')]
+    exact h_mgf t
+
+lemma _root_.ProbabilityTheory.Kernel.HasSubgaussianMGF_congr {Y : Ω → ℝ} (h : X =ᵐ[κ ∘ₘ ν] Y) :
+    HasSubgaussianMGF X c κ ν ↔ HasSubgaussianMGF Y c κ ν :=
+  ⟨fun hX ↦ congr hX h, fun hY ↦ congr hY (ae_eq_symm h)⟩
+
+section ChernoffBound
+
+lemma measure_ge_le_exp_add [IsFiniteKernel κ] (h : HasSubgaussianMGF X c κ ν) (ε : ℝ) :
```

Comment:
Do you really need the kernel to be finite here? The subgaussian assumption implies that, almost everywhere, the total mass of `κ ω'` is at most 1 (take `t = 0`), is that enough for your needs?

Reply:
The sub-Gaussian assumption at 0 gives me that a.e. `(κ ω' Set.univ).toReal ≤ 1`, but due to the `toReal` that's also true if the mass is infinite.

Reply:
true, but `ae_integrable_exp_mul` with `t = 0` tells you that the constant 1 is integrable, so this should exclude the infinite mass case, no?

Reply:
That one indeed works.